File: backend-node/utils/encryption.js
import crypto from "crypto"
 
const ALGORITHM  = "aes-256-gcm"
const KEY_LENGTH = 32 
const IV_LENGTH  = 12  
const TAG_LENGTH = 16 

// Key Derivation 
// Derives a unique 32-byte key per user using HMAC-SHA256
// Same derivation logic used in FastAPI encryption.py
 
function deriveKey(userId) {
    const secret = process.env.ENCRYPTION_SECRET
    if (!secret) throw new Error("ENCRYPTION_SECRET not set in .env")
 
    return crypto
        .createHmac("sha256", secret)
        .update(userId.toString())
        .digest()  
}
 
// Encrypt
// Returns base64 string: iv(12 bytes) + authTag(16 bytes) + ciphertext
// Returns "" for empty/null input — no point encrypting empty strings
 
export function encrypt(plaintext, userId) {
    if (!plaintext || plaintext === "") return ""
 
    try {
        const key = deriveKey(userId)
        const iv  = crypto.randomBytes(IV_LENGTH)
 
        const cipher = crypto.createCipheriv(ALGORITHM, key, iv)
 
        const encrypted = Buffer.concat([
            cipher.update(plaintext, "utf8"),
            cipher.final()
        ])
 
        const authTag = cipher.getAuthTag()
 
        // Layout: iv(12) + tag(16) + ciphertext
        // FastAPI decrypt expects the same layout
        const combined = Buffer.concat([iv, authTag, encrypted])
        return combined.toString("base64")
 
    } catch (error) {
        console.log("Encryption error:", error)
        throw error
    }
}

File: backend-fastapi/utils/encryption.py
import os
import hmac
import hashlib
import base64

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

IV_LENGTH  = 12
TAG_LENGTH = 16 

# Key Derivation
# Must produce identical output to Node.js deriveKey()
# Node.js: crypto.createHmac("sha256", secret).update(userId).digest()
# Python:  hmac.new(secret, userId, sha256).digest()

def derive_key(user_id: str) -> bytes:
    secret = os.getenv("ENCRYPTION_SECRET")
    if not secret:
        raise ValueError("ENCRYPTION_SECRET not set in .env")

    return hmac.new(
        secret.encode("utf-8"),
        user_id.encode("utf-8"),
        hashlib.sha256
    ).digest()  # 32 bytes — matches AES-256 key size

# Decrypt
# Expects base64 string with layout: iv(12) + tag(16) + ciphertext
# This is the exact layout Node.js encrypt() produces
# Returns "" on empty input or decryption failure (so tool never crashes)

def decrypt(ciphertext: str, user_id: str) -> str:
    if not ciphertext or ciphertext == "":
        return ""

    try:
        key = derive_key(user_id)
        combined = base64.b64decode(ciphertext)

        # Extract components — same layout as Node.js
        iv = combined[:IV_LENGTH]
        tag = combined[IV_LENGTH : IV_LENGTH + TAG_LENGTH]
        encrypted = combined[IV_LENGTH + TAG_LENGTH :]

        aesgcm = AESGCM(key)

        # AESGCM.decrypt() expects ciphertext + tag concatenated as one arg
        decrypted = aesgcm.decrypt(iv, encrypted + tag, None)

        return decrypted.decode("utf-8")

    except Exception as e:
        print(f"Decryption error for user {user_id}: {e}")
        return ""  # return empty rather than crash the tool